server/scsynth/SC_PortAudio.cpp

```diff
@@ -303,25 +468,47 @@ bool SC_PortAudioDriver::DriverSetup(int* outNumSamples, double* outSampleRate)
     fprintf(stdout, "\nDevice options:\n");
     for (int i = 0; i < numDevices; i++) {
         pdi = Pa_GetDeviceInfo(i);
-        apiInfo = Pa_GetHostApiInfo(pdi->hostApi);
-        fprintf(stdout, "  - %s : %s   (device #%d with %d ins %d outs)\n", apiInfo->name, pdi->name, i,
+        fprintf(stdout, "  - %s   (device #%d with %d ins %d outs)\n", GetPaDeviceName(i).c_str(), i,
                 pdi->maxInputChannels, pdi->maxOutputChannels);
     }
 
     mDeviceInOut[0] = paNoDevice;
     mDeviceInOut[1] = paNoDevice;
     if (mWorld->hw->mInDeviceName)
-        GetPaDeviceFromName(mWorld->hw->mInDeviceName, mDeviceInOut);
-    if (mDeviceInOut[0] == paNoDevice)
-        mDeviceInOut[0] = Pa_GetDefaultInputDevice();
-    if (mDeviceInOut[1] == paNoDevice)
-        mDeviceInOut[1] = Pa_GetDefaultOutputDevice();
+        GetPaDeviceFromName(mWorld->hw->mInDeviceName, &mDeviceInOut[0], IOType::Input);
+    if (mWorld->hw->mOutDeviceName)
+        GetPaDeviceFromName(mWorld->hw->mOutDeviceName, &mDeviceInOut[1], IOType::Output);
+
+    // report requested devices
+    fprintf(stdout, "\nRequested devices:\n");
+    if (mWorld->mNumInputs) {
+        fprintf(stdout, "  In (matching device %sfound):\n  - %s\n", (mDeviceInOut[0] == paNoDevice ? "NOT " : ""),
+                mWorld->hw->mInDeviceName);
+    }
+    if (mWorld->mNumOutputs) {
+        fprintf(stdout, "  Out (matching device %sfound):\n  - %s\n", (mDeviceInOut[1] == paNoDevice ? "NOT " : ""),
+                mWorld->hw->mOutDeviceName);
+    }
+
+    fprintf(stdout, "\n");
+    paerror = CheckPaDevices(&mDeviceInOut[0], &mDeviceInOut[1], mWorld->mNumInputs, mWorld->mNumOutputs,
+                             mPreferredSampleRate);
+
+    // if we got an error from CheckPaDevices, stop here
+    if (paerror != paNoError) {
+        PRINT_PORTAUDIO_ERROR(Pa_OpenStream, paerror);
+        return paerror == paNoError;
+    }
 
     *outNumSamples = mWorld->mBufLength;
     if (mPreferredSampleRate)
         *outSampleRate = mPreferredSampleRate;
-    else
-        *outSampleRate = 44100.;
+    else {
+        if (mDeviceInOut[0] != paNoDevice)
+            *outSampleRate = Pa_GetDeviceInfo(mDeviceInOut[0])->defaultSampleRate;
+        if (mDeviceInOut[1] != paNoDevice)
+            *outSampleRate = Pa_GetDeviceInfo(mDeviceInOut[1])->defaultSampleRate;
```

Comment:
is this right?

Reply:
It is... intended - outSampleRate is _the_ sample rate.

In case we only have input device (i.e. no output), then query input for sampleRate. If we have both input and output, or just the output, then use output device to get the sample rate.

Reply:
ok, just confirming. a clearer way to write this is: 

```cpp
    if (mPreferredSampleRate)
        *outSampleRate = mPreferredSampleRate;
    else if (mDeviceInOut[1] != paNoDevice)
        *outSampleRate = Pa_GetDeviceInfo(mDeviceInOut[1])->defaultSampleRate;
    else if (mDeviceInOut[0] != paNoDevice)
        *outSampleRate = Pa_GetDeviceInfo(mDeviceInOut[0])->defaultSampleRate;
```

btw, won't you be leaving `outSampleRate` unset if both devices are no-device? IIUC it will then be used in that call to [`Pa_OpenDefaultStream`](https://github.com/dyfer/supercollider/blob/topic/scsynth-portaudio-input-output-devices/server/scsynth/SC_PortAudio.cpp#L585) you mentioned before; you probably want to have set this to some reasonable value for that call.